src/s_stuff.h

```diff
@@ -10,6 +10,8 @@ in future releases.  The public (stable) API is in m_pd.h. */
 
 /* in s_path.c */
 
+#include <sys/socket.h> // struct sockaddr, socklen_t
```

Comment:
This obviously doesn't work on Windows :-)

Actually, we only need a definition for `socklen_t`, since `struct sockaddr *` can be an incomplete type. However, we don't want to include `winsock2.h`, as it pollutes the namespace with Windows stuff.

We *could* do `typedef int socklen_t`, but we already do that in `s_net.h` and I think C doesn't permit duplicate typedefs (C++ does), so we would get an error in `s_inter.c` (which includes both headers).

I see two solutions:

1) Don't include any socket headers in `s_stuff.h` at all and simply use an `int` or `size_t` for the sockaddr length arguments in `sys_sendto` resp. `rbskt_recvfrom`
2) For Windows, define `socklen_t` like this in both `s_stuff.h` and `s_net.h`:
  ```
    #ifndef socklen_t
      #define socklen_t int
    #endif
   ```

@danomatika What do you think?

Reply:
@danomatika 
> I’d prefer option 1 with size_t.

The issue with `size_t` is that `socklen_t` is defined (at least on Linux x64 and on ARMv7 and macos), as a 32 bit `int`.`size_t` on the other hand would be 64bit on a 64 bit platform. This could be taken care of within the function as needed, but as these functions aim to have the same signature, return and `errno` behaviour as `send()`/`sendto()`, it should probably be an `int`?

Reply:
In general, I prefer to move things towards `size_t` where possible, which balancing the fact that adding *anything* to the API will probably be written into *stone* so we want to get this right the first time.

`size_t` was added to C for generic "size" and "len" variable usage which unsigned values. As far as I know, `size_t` was added to C after the Berkley Sockets API was written.

In any case, if the value is a `int` type and there is no major move towards expanding that, `int` is probably fine. If there is a chance we would need to be able to handle a type other than 32 bit, then it should be something like `size_t`.

Reply:
I think @guiliomoro raised an important point that the functions are supposed to have the same signature as their socket API counterparts. For `socklen_t` vs `size_t` arguments it's not a problem, but `socklen_t *` and `size_t *` are mostly incompatible.

In practice that's probably not a big deal, as client code already has to call different functions anyway, so they might simply use an additional local variable for the address length:
```
struct sockaddr_storage ss;
socklen_t addrlen;
if (use_ringbuffer) {
    size_t len = sizeof(ss);
    int result = rbskt_recvfrom(rb, buffer, bufsize, 0, (struct sockaddr *)&ss, &len))
    if (result < 0) {
        // error handling
    }
    addrlen = len;
} else {
    addrlen = sizeof(ss);
    int result = recvfrom(fd, buffer, bufsize, 0, (struct sockaddr *)&ss, &addrlen))
    if (result < 0) {
        // error handling
    }
}
```

> In any case, if the value is a int type and there is no major move towards expanding that, int is probably fine.

`socklen_t` can be signed or unsigned and has to be *at least* 32 bit. The only way to get it right is to include `sys/socket.h`. If we don't want to do that, we should just go for `size_t`.

Reply:
> In practice, using size_t is not a big deal, as client code already has to call different functions anyway, so they might simply use an additional local variable for the address length:

this seems error-prone

here's an idea:

```
EXTERN ssize_t sys_sendto(int sockfd, const void *buf, size_t len, int flags, const struct sockaddr *dest_addr, size_t dest_len);
EXTERN int rbskt_recvfrom(t_rbskt* rbskt, void* buf, size_t buflen, int nothing, struct sockaddr *address, socklen_t* address_len);
```

note the `size_t dest_len` vs `socklen_t* address_len`:
- when it's passed by value it's a `size_t`; if on the system at hand `socklen_t` is different from `size_t`, the compiler will take care of that.
- when it's a pointer, the header file doesn't care and we can use an opaque pointer. Again, the compiler will take care of that when building the implementation file.

Reply:
> this seems error-prone

Why is this error prone? If you pass the wrong pointer type, you get a compilation error. It can't get any safer than that :-)

> when it's a pointer, the header file doesn't care and we can use an opaque pointer. Again, the compiler will take care of that when building the implementation file.

How would you make a forward declaration of `socklen_t`?

---

If we really want make the function signatures match the socket API, I don't really see another way than method 2 in https://github.com/pure-data/pure-data/pull/1261#discussion_r563285436

Reply:
> Why is this error prone? If you pass the wrong pointer type, you get a compilation error. It can't get any safer than that :-)

I think you'll only get a compiler warning, not error (at least with the current build settings of Pd) and there are some warnings already when building Pd (or better `portmidi`, which gets built as part of it). Either way, I wouldn't want to rely on a compiler warning.

> How would you make a forward declaration of socklen_t?

I was thinking of using `void*` instead of `socklen_t*`, which would compile but make it unsafer than your previous solution which I was trying to improve upon. So, a bad idea from my side.

Reply:
> I think you'll only get a compiler warning, not error

You're right! By default, C compilers would only issue a warning. Only in C++ it is always an error.